File: src/chain-observers/filters/custom-operation-filter.ts
import { OperationClassifier } from "../classifiers";
import type { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import type { TFilterEvaluationContext } from "../factories/data-evaluation-context";
import { FilterBase } from "./filter-base";

export class CustomOperationFilter extends FilterBase {
  public constructor(
    ids: Array<string | number>
  ) {
    super();

    this.ids = new Set(ids);
  }

  private readonly ids = new Set<string | number>();

  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [
      OperationClassifier
    ];
  }

  public async match(data: TFilterEvaluationContext): Promise<boolean> {
    const { operationsPerType } = await data.get(OperationClassifier);

    for(const { operation } of (operationsPerType.custom_json_operation ?? []))
      if (this.ids.has(operation.id))
        return true;

    return false;
  }
}

File: python/workerbee/chain_observers/filters/custom_operation_filter.py
"""CustomOperationFilter — matches custom_json_operation with specified ids.

Mirrors src/chain-observers/filters/custom-operation-filter.ts.

Uses OperationClassifier.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .._ordered_set import OrderedSet
from .filter_base import FilterBase

if TYPE_CHECKING:
    from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
    from ..factories.data_evaluation_context import DataEvaluationContext


class CustomOperationFilter(FilterBase):
    def __init__(self, ids: list[str | int]) -> None:
        super().__init__()
        self._ids: OrderedSet[str | int] = OrderedSet(ids)

    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        from ..classifiers.operation_classifier import OperationClassifier

        return [OperationClassifier]

    async def match(self, data: DataEvaluationContext) -> bool:
        from ..classifiers.operation_classifier import OperationClassifier

        result = await data.get(OperationClassifier)
        operations_per_type = result["operations_per_type"]

        custom_json_ops = operations_per_type.get("custom_json_operation", [])
        return any(entry["operation"]["id"] in self._ids for entry in custom_json_ops)
